src/Watcher.cs

```diff
@@ -0,0 +1,100 @@
+﻿using System;
+using System.IO;
+using System.Runtime.Serialization;
+using System.Threading;
+using System.Threading.Tasks;
+using k8s.Exceptions;
+using Microsoft.Rest;
+using Microsoft.Rest.Serialization;
+
+namespace k8s
+{
+    public enum WatchEventType
+    {
+        [EnumMember(Value = "ADDED")] Added,
+
+        [EnumMember(Value = "MODIFIED")] Modified,
+
+        [EnumMember(Value = "DELETED")] Deleted,
+
+        [EnumMember(Value = "ERROR")] Error
+    }
+
+    public class Watcher<T> : IDisposable
+    {
+        private readonly StreamReader _streamReader;
+        private readonly CancellationTokenSource _cts;
+
+        public Watcher(StreamReader streamReader, Action<WatchEventType, T> onEvent, Action<Exception> onError)
+        {
+            _streamReader = streamReader;
+            OnEvent += onEvent;
+            OnError += onError;
+
+            _cts = new CancellationTokenSource();
+
+            var token = _cts.Token;
+
+            Task.Run(async () =>
+            {
+                while (!streamReader.EndOfStream)
+                {
+                    if (token.IsCancellationRequested)
+                    {
+                        return;
+                    }
+
+                    try
+                    {
+                        var line = await streamReader.ReadLineAsync();
+                        var @event = SafeJsonConvert.DeserializeObject<WatchEvent>(line);
+
+                        OnEvent?.Invoke(@event.Type, @event.Object);
+                    }
+                    catch (Exception e)

```

Comment:
Isn't catching all exceptions an anti-pattern? Maybe just the parsing and IO exceptions?

Reply:
hi please have a look at my testcases which would better illustrate what I want to do.

* TestSuriveBadLine: the watch object will still alive when json parse error or exception from onevent. the `onerror` will be triggered
* TestWatchServerDisconnect: the watch object stops as of transport level error. the `onerror` will be triggered